pravega/flink-connectors
# [issue-266] Writer side: Implement watermark with Pravega

**Change log description**
- Add API support for watermark in normal transactional and non-transactional writer
- Add tests for Pravega watermark writer and Table API.

**Purpose of the change**
- Add Pravega watermarking support for writer. 

**What the code does**
Provide a `enableWatermark()` API, when this switch is on:
 
- For non-transactional writer
  - In `write()`, we track Flink current watermark, if event time passes the watermark, we call `noteTime(t)`

- For transactional writer
   - In `write()`, we store Flink current watermark in the writer state. 
   - In `snapshotState()`, store `currentWatermark` into the `PendingTransaction` state.  
   - In `notifyCheckpointComplete()` and `restoreState()`, if transaction state has watermark, commit with watermark.

- For Table API
  - Implement with underlying `DataStreamSink`

**How to verify it**
Three test cases:
For non-transactional writer, write Integer events with watermark enabled. Then read it to see if the `TimeWindows` progress. (Now this case failed blocked by pravega#4228)
For transactional writer, ensure all data written exactly once.
For Table API, check the Table Sink initialization and validation passes.

`./gradlew clean build` passes.
